torch/csrc/distributed/c10d/init.cpp

```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
           "wait",
           &::c10d::ProcessGroup::Work::wait,
           py::arg("timeout") = kNoTimeout,
-          py::call_guard<py::gil_scoped_release>());
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "get_future",
+          [](::c10d::ProcessGroup::Work& work)
+              -> std::shared_ptr<jit::PythonFutureWrapper> {
+            return std::make_shared<jit::PythonFutureWrapper>(work.getFuture());
+          },
+          py::call_guard<py::gil_scoped_release>(),
+          R"(
+            ``get_future`` retrieves a future associated with the completion of
+            ``c10d.ProcessGroup.work``. As an example, a future object can be
+            retrieved by `fut = dist.allreduce(tensors).get_future()`.
+
+            In the example above, if `dist.allreduce` work was done on GPU,
+            `fut.wait()` would return after synchronizing the correct GPU streams
+            to ensure we can have async CUDA execution and it does not wait for
+            the entire operation to complete on GPU. If NCCL_BLOCKING_WAIT is enabled,
+            in that case, it would wait for the entire operation to complete before
+            returning. In addition, if a callback function was added by `fut.then()`,
+            it will synchronize the appropriate streams and invoke the callback inline.
+
+            Note that `fut.done()` actually returns if the work was completed on the GPU.
+
+            .. warning ::
```

Comment:
Add one more warning to highlight how this is different from `torch.futures.Future`.

Reply:
I updated the description. I'm not sure about the differences between `toch_C.Future` and `torch.futures.Future`. I asked that question in a separate thread.

Reply:
This is actually not about the difference between `torch.futures.Future` and `torch._C.Future`, instead, we want to distinguish `torch.futures.Future` from the Future type returned by `get_future`.  We are just using the `torch._C.Future` type to infer that "this is an internal type, please read the limitations and use it with caution".